### DISCUSSION TITLE
[Help Request] - Identical cluster/command combo does different things - set/get/write

### DISCUSSION THREAD

**Discussion by rechichidaniel** [Koenkk/zigbee2mqtt#16282] [General]
Hi. Can someone please help me with this? I've been stuck for days.

# What I'm Trying To Do
Construct MQTT/JSON commands to write or read attributes.

# What I've Attempted
Many things but I'll use setting the power_on_behavior of my light as an example.

Using the Zigbee2MQTT front end, I can set the power_on_behavior to previous from the Exposes menu.
I can see in my network traffic that my Zigbee coordinator sends the `Write Attributes (0x02)` command on the `On/Off (0x0006)` cluster, attaching the `StartUpOnOff (0x4003)` attribute and `Previous (0xff)` value (is this field called value?).

The light then responds to the coordinator with the `Write Attributes Response (0x04)` command on the `On/Off (0x0006)` cluster with the `Success (0x00)` status.

I thought I knew from all that what information I needed to build the MQTT message but something is definitely missing.
When I try to send the `Write Attributes (0x02)` command on the `On/Off (0x0006)` cluster, I'm unintentionally sending the `Toggle (0x02)` command on the `On/Off (0x0006)` cluster. Admittedly, I haven't figured out how to attach the attributes payload yet, but nothing I've tried putting in the json payload changes the command sent.
The light responds to this toggle with the `Default Response (0x0b)` command on the `On/Off (0x0006)` cluster with the `Success (0x00)` status.

# Raw Network Traffic
If you also want the responses, please let me know.
## The `Write Attributes (0x02)` command I want to send.
(Sorry, I have no idea what language to use in this code block for WireShark traffic, but javascript looks readable if you can ignore some of the inconsistencies)
```javascript
Frame 1: 50 bytes on wire (400 bits), 50 bytes captured (400 bits) on interface COM7, id 0
    (...)
IEEE 802.15.4 Data, Dst: 0x1b6d, Src: 0x0000
    (...)
ZigBee Network Layer Data, Dst: 0x1b6d, Src: 0x0000
    (...)
ZigBee Application Support Layer Data, Dst Endpt: 1, Src Endpt: 1
    Frame Control Field: Data (0x00)
        .... ..00 = Frame Type: Data (0x0)
        .... 00.. = Delivery Mode: Unicast (0x0)
        ..0. .... = Security: False
        .0.. .... = Acknowledgement Request: False
        0... .... = Extended Header: False
    Destination Endpoint: 1
    Cluster: On/Off (0x0006)
    Profile: Home Automation (0x0104)
    Source Endpoint: 1
    Counter: 111
ZigBee Cluster Library Frame, Command: Write Attributes, Seq: 73
    Frame Control Field: Profile-wide (0x10)
        .... ..00 = Frame Type: Profile-wide (0x0)
        .... .0.. = Manufacturer Specific: False
        .... 0... = Direction: Client to Server
        ...1 .... = Disable Default Response: True
    Sequence Number: 73
    Command: Write Attributes (0x02)
    Attribute Field
        Attribute: StartUpOnOff (0x4003)
        Data Type: 8-Bit Enumeration (0x30)
        Startup On Off: Set the OnOff attribute to its previous value (0xff)
```
## The `Toggle (0x02)` command I unintentionally send.
```javascript
Frame 3: 46 bytes on wire (368 bits), 46 bytes captured (368 bits) on interface COM7, id 0
    (...)
IEEE 802.15.4 Data, Dst: 0x1b6d, Src: 0x0000
    (...)
ZigBee Network Layer Data, Dst: 0x1b6d, Src: 0x0000
    (...)
ZigBee Application Support Layer Data, Dst Endpt: 1, Src Endpt: 1
    Frame Control Field: Data (0x00)
        .... ..00 = Frame Type: Data (0x0)
        .... 00.. = Delivery Mode: Unicast (0x0)
        ..0. .... = Security: False
        .0.. .... = Acknowledgement Request: False
        0... .... = Extended Header: False
    Destination Endpoint: 1
    Cluster: On/Off (0x0006)
    Profile: Home Automation (0x0104)
    Source Endpoint: 1
    Counter: 112
ZigBee Cluster Library Frame
    Frame Control Field: Cluster-specific (0x01)
        .... ..01 = Frame Type: Cluster-specific (0x1)
        .... .0.. = Manufacturer Specific: False
        .... 0... = Direction: Client to Server
        ...0 .... = Disable Default Response: False
    Sequence Number: 74
    Command: Toggle (0x02)
```
## How I'm unintentionally sending the `Toggle (0x02)` command.
Topic:
```html
zigbee2mqtt/Z2M Test Light/set
```
Payload
```json
{
    "command": {
        "cluster": 6,
        "command": 2,
        "payload": {
            "attributeid": 16387,
            "value": 255
        }
    }
}
```
I've also tried changing the payload attributes and also removing the payload entirely.

**Comment by rechichidaniel** [comment]
I've also tried reading the power_on_behavior by changing the MQTT topic from `/set` to `/get` but I get an error in the Z2M logs saying "_No converter available for 'get' 'command' ([object Object])_"
```
zigbee2mqtt/Z2M Test Light/get
```
When clicking the refresh button next to power_on_behavior in the lights Exposes menu, the coordinator sends the `Read Attributes (0x00)` command on the `On/Off (0x0006)` cluster with the `StartUpOnOff (0x4003)` attribute. This is what I was trying to copy when I changed my MQTT topic from `/set` to `/get`.
```javascript
Frame 5: 48 bytes on wire (384 bits), 48 bytes captured (384 bits) on interface COM7, id 0
    (...)
IEEE 802.15.4 Data, Dst: 0x1b6d, Src: 0x0000
    (...)
ZigBee Network Layer Data, Dst: 0x1b6d, Src: 0x0000
    (...)
ZigBee Application Support Layer Data, Dst Endpt: 1, Src Endpt: 1
    Frame Control Field: Data (0x00)
        .... ..00 = Frame Type: Data (0x0)
        .... 00.. = Delivery Mode: Unicast (0x0)
        ..0. .... = Security: False
        .0.. .... = Acknowledgement Request: False
        0... .... = Extended Header: False
    Destination Endpoint: 1
    Cluster: On/Off (0x0006)
    Profile: Home Automation (0x0104)
    Source Endpoint: 1
    Counter: 113
ZigBee Cluster Library Frame, Command: Read Attributes, Seq: 75
    Frame Control Field: Profile-wide (0x10)
        .... ..00 = Frame Type: Profile-wide (0x0)
        .... .0.. = Manufacturer Specific: False
        .... 0... = Direction: Client to Server
        ...1 .... = Disable Default Response: True
    Sequence Number: 75
    Command: Read Attributes (0x00)
    Attribute: StartUpOnOff (0x4003)
```

**Reply by roadsnail** [reply]
Hi, don't know if this will help much, but I wanted to use the frontend 'Dev console' to write out 4 attributes to my heating controller in one zigbee cluster library frame. I had previously sniffed the Dev console 'write' and discovered using this API worked in my user case.

So this is what I did to arrive at the payload generated by the 'write' API which I was then able to use in my node-red flow...

![2023-02-01 12_40_00-](https://user-images.githubusercontent.com/24318993/216062064-3ed169ea-17d8-4586-95ce-a12e9ae54581.png)

The above shows me using the browser inspect of the Dev Console 'write' button to see that an API is called 3 times. A read, write followed by a further read. 

![2023-02-01 12_41_33-](https://user-images.githubusercontent.com/24318993/216062762-5b6fca0a-54b6-42af-b143-ddb09082230d.png)

Drilling down to the write API, I then copied the object which I was then able to use in my node-red flow to create what I wanted. ie the four attributes sent to the SLR2 controller in one zigbee frame.

Maybe this will help a little in your user case?

**Reply by rechichidaniel** [reply]
Thanks roadsnail,
That might help but what did your final Node Red json look like? I'm stuck on the last step where I move that to a Node Red MQTT node.

**Comment by sjorge** [comment]
I'm very confused as to what you are trying to do.

Are you trying to read/write custom payloads? Or do you just want to set/get the exposed properties you see in the frontend?
The former is possible but it's very annoying to do and honestly only needed when trying to add support for a new device.

The later is easy, although you can't read everything, only things that have the blue refresh icon.

`light/office/desk/bulb` is my friendly name in the example below.

Since you mentioned power_on_behavior, you can set it like this:
```
mosquitto_pub -t zigbee2mqtt/light/office/desk/bulb/set -m '{"power_on_behavior": "off"}'
```
Or get the value like this:
```
mosquitto_pub -t zigbee2mqtt/light/office/desk/bulb/get -m '{"power_on_behavior": ""}
```

You need to listen on `zigbee2mqtt/light/office/desk/bulb` to see the result once it is published after a get.

**Reply by rechichidaniel** [reply]
Thanks Jorge,

To clear up some confusion, I'm stuck trying to send the `write attributes (0x02)` command because when I try, I end up sending the `Command: Toggle (0x02)` command instead.

I'd like to be able to programmatically send the `write attributes (0x02)` command because I kind of am trying to add a new device. I'm testing an existing device that is a bit buggy and I need to be able to send commands that aren't added to the existing Z2M extend command for that particular device. I'm also developing my own hardware using the Nordic nRF52 series which I will need to be able to add converters for and I'd like to use Z2M to help me develop.

`power_on_behavior` was just an example that @Koenkk helped me with in [issue:16283](https://github.com/Koenkk/zigbee2mqtt/issues/16283). But that has a different solution that I won't be able to use for other things.

**Reply by sjorge** [reply]
I don't think that is possible without modifications to zigbee-herdsman (I think). You can write 'known' cluster.attribute pairs using the write converter.

```
 mosquitto_pub -t zigbee2mqtt/thermostat/home/set/write -m '{"cluster": "hvacThermostat", "payload": {"systemMode": 0}}'
```

But you can't do 
```
mosquitto_pub -t zigbee2mqtt/thermostat/home/set/write -m '{"cluster": 513, "payload": {28: 0}}'
```

The later fails in `entity.write()` when called by the ['write' converter](https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/converters/toZigbee.js#L39-L49) 

I don't 100% understand why it is failing as according to https://github.com/Koenkk/zigbee-herdsman/blob/1f863593b9e39237a5fc3b4ee540e3f64a28d0e7/src/controller/model/endpoint.ts#L361 where it ultimately ends up it should accept numbers :/

```
debug 2023-02-12 16:03:22: Error: Cluster with name 'undefined' does not exist
    at Object.getCluster (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/zcl/utils.ts:61:19)
    at Endpoint.write (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:364:35)
    at Object.convertSet (/opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/converters/toZigbee.js:46:26)
    at Publish.onMQTTMessage (/opt/zigbee2mqtt/lib/extension/publish.ts:246:52)
    at EventEmitter.emit (node:events:525:35)
    at EventBus.emitMQTTMessage (/opt/zigbee2mqtt/lib/eventBus.ts:109:22)
    at MQTT.onMessage (/opt/zigbee2mqtt/lib/mqtt.ts:140:27)
    at MqttClient.emit (node:events:513:28)
    at MqttClient._handlePublish (/opt/zigbee2mqtt/node_modules/mqtt/lib/client.js:1547:12)
    at MqttClient._handlePacket (/opt/zigbee2mqtt/node_modules/mqtt/lib/client.js:535:12)
```
